src/main/java/com/microsoft/azure/proton/transport/proxy/impl/ProxyImpl.java

```diff
@@ -274,5 +307,109 @@ public void close_head() {
             headClosed = true;
             underlyingOutput.close_head();
         }
+
+        private void getChallengeQuestionHeaders(String line, Map<String, String> challengeQuestionValues) {
+            String context =  line.substring(PROXY_AUTH_DIGEST.length());
+            String[] headerValues = context.split(",");
+
+            for (String headerValue : headerValues) {
+                if (headerValue.contains("=")) {
+                    String key = headerValue.substring(0, headerValue.indexOf("="));
+                    String value = headerValue.substring(headerValue.indexOf("=") + 1);
+                    challengeQuestionValues.put(key.trim(), value.replaceAll("\"", "").trim());
+                }
+            }
+        }
+
+        private void computeBasicAuthHeader(){
+            final PasswordAuthentication authentication = Authenticator.requestPasswordAuthentication(
+                    "",
+                    null,
+                    0,
+                    "https",
+                    "Event Hubs client websocket proxy support",
+                    "basic",
+                    null,
+                    Authenticator.RequestorType.PROXY);
+            if (authentication == null) return;
+
+            final String proxyUserName = authentication.getUserName();
+            final String proxyPassword = authentication.getPassword() != null
+                    ? new String(authentication.getPassword())
+                    : null;
+            if (isNullOrEmpty(proxyUserName) || isNullOrEmpty(proxyPassword))  return;
+
+            final String usernamePasswordPair = proxyUserName + ":" + proxyPassword;
+            if (headers == null)
+                headers = new HashMap<String, String>();
+            headers.put(
+                    "Proxy-Authorization",
+                    "Basic " + Base64.getEncoder().encodeToString(usernamePasswordPair.getBytes()));
+        }
+
+        private void computeDigestAuthHeader(Map<String, String> challengeQuestionValues) {
+            String uri = host;
+            PasswordAuthentication passwordAuthentication = Authenticator.requestPasswordAuthentication(
+                    "",
+                    null,
+                    0,
+                    "https",
+                    "Event Hubs client websocket proxy support",
+                    "digest",
+                    null,
+                    Authenticator.RequestorType.PROXY);
+
+            String username = passwordAuthentication.getUserName();
+            String password =  passwordAuthentication.getPassword() != null
+                    ? new String(passwordAuthentication.getPassword())
+                    : null;
+
+            String digestValue;
+            try {
+                SecureRandom random = SecureRandom.getInstance("SHA1PRNG");
+                random.setSeed(System.currentTimeMillis());
+                byte[] nonceBytes = new byte[16];

```

Comment:
I cannot find that nonceBytes is ever used. It looks like these four lines can be removed.

Reply:
**nonceBytes** is used as a container to host the random-generated bytes, which converts to a string containing a lexical representation of hexBinary, **cnonce**.  

_secureRandom.nextBytes(cnonceBytes);
String cnonce = DatatypeConverter.printHexBinary(cnonceBytes).toLowerCase();_

Reply:
That's cnonceBytes, which is a separate array from nonceBytes. There's a similar duplication with random/secureRandom.

Reply:
Yeah, thanks.  Just delete these 4 lines with a new commit.